Bad bytecode generated for a partial function.

When running the example below, it fails with `java.lang.VerifyError` [Full stack trace](https://github.com/lampepfl/dotty/files/983955/stacktrace.txt)
```scala
class Main {
  val buzz = Some(Bees.Bee("buzz")).collect {
    case Bees.Bee(value) => value
  }
}

object Main {
  def main(args: Array[String]): Unit = {
    new Main
  }
}

object Bees {
  case class Bee(value: String)
}
```

Reproducible in fresh `lampepfl/dotty.g8`
OpenJDK 8 on Ubuntu
```
$ java -version
openjdk version "1.8.0_121"
OpenJDK Runtime Environment (build 1.8.0_121-8u121-b13-0ubuntu1.16.04.2-b13)
OpenJDK 64-Bit Server VM (build 25.121-b13, mixed mode)
```

A simpler case that exhibits a similar issue is:
```scala
class Bees {
  def f: PartialFunction[Bee, Unit] = {  case Bee(_) => ""  }

  f(new Bee("buzz"))

  case class Bee(value: String)
  // object Bee // With this it works
}

object Test {
  def main(args: Array[String]): Unit = {
    new Bees
  }
}
```
fails with 
```scala
Caused by: java.lang.NullPointerException
	at Bees.Bees$f$$f$$anonfun$1$1(i2396.scala:6)
	at Bees$$anonfun$1.apply(i2396.scala:6)
	at Bees$$anonfun$1.apply(i2396.scala:6)
	at Bees.<init>(i2396.scala:9)
	at Test$.main(i2396.scala:17)
	at Test.main(i2396.scala)
```
where the bytecode has
```java
public class Bees {
    public final Bees.Bee$ Bee$lzy1;

    public Bees() {
        this.f().apply(new Bees.Bee(this, "buzz"));
        this.Bee$lzy1 = new Bees.Bee$(this); // should be lazy
    }

    public final Bees.Bee$ Bee() { return this.Bee$lzy1; }  // should be initialized in here if needed
   ...
}
```

--- a/compiler/src/dotty/tools/dotc/transform/ElimStaticThis.scala
+++ b/compiler/src/dotty/tools/dotc/transform/ElimStaticThis.scala
@@ -27,10 +27,10 @@ class ElimStaticThis extends MiniPhaseTransform {
   override def transformIdent(tree: tpd.Ident)(implicit ctx: Context, info: TransformerInfo): tpd.Tree = {
     if (ctx.owner.enclosingMethod.is(JavaStatic)) {
       tree.tpe match {
-        case TermRef(thiz: ThisType, _) if thiz.cls.is(ModuleClass) =>
+        case TermRef(thiz: ThisType, _) if thiz.cls.is(ModuleClass, JavaDefined) =>
           ref(thiz.cls.sourceModule).select(tree.symbol)
         case TermRef(thiz: ThisType, _) =>
-          assert(tree.symbol.is(Flags.JavaStatic))
+          assert(tree.symbol.is(Flags.JavaStatic) || thiz.cls.is(JavaDefined))
           tree
         case _ => tree
       }

--- a/tests/run/i2396.scala
+++ b/tests/run/i2396.scala
@@ -1,0 +1,15 @@
+class B {
+  val buzz = Some(Bees.Bee("buzz")).collect {
+    case Bees.Bee(value) => value
+  }
+}
+
+object Test {
+  def main(args: Array[String]): Unit = {
+    new B
+  }
+}
+
+object Bees {
+  case class Bee(value: String)
+}

--- a/tests/run/i2396c.scala
+++ b/tests/run/i2396c.scala
@@ -1,0 +1,15 @@
+class Bees {
+  import Test._
+
+  def f: PartialFunction[Bee, Unit] = {
+     case Test.Bee(_) => ""
+ //    case Bee(_) => ""  // This one works
+  }
+
+  f(new Bee("buzz"))
+}
+
+object Test {
+  case class Bee(value: String)
+  def main(args: Array[String]): Unit = new Bees
+}

